library/Fission/IPFS/Peer.hs

```diff
@@ -54,5 +62,45 @@ connect peer@(Peer peerID) = IPFSProc.run ["swarm", "connect"] (UTF8.textToLazyB
   (ExitFailure _ , _, _) -> Left $ CannotConnect peer
   (ExitSuccess   , _, _) -> Right ()
 
+peerAddressRe :: Regex
+peerAddressRe = mkRegex "^/ip[46]/([a-zA-Z0-9.:]*)/"
+
+extractIPfromPeerAddress :: String -> Maybe String
+extractIPfromPeerAddress peer = peer
+  & matchRegex peerAddressRe
+  <&> Data.List.head
+
+isExternalIPv4 :: Text -> Bool
+isExternalIPv4 ip = ip
+  & T.unpack
+  & extractIPfromPeerAddress
+  & maybe "" T.pack
+  & Net.IPv4.decode
+  <&> Net.IPv4.reserved
+  & maybe False not
```

Comment:
Trying to get this to just pipe the whole way down is an interesting approach — I can't say that I even dislike it. It's Haskell with an Elixir accent, which I'm clearly not against at the author of Witchcraft 🔮

For science, below is more typical phrasing. The Haskell community really focuses on semantics (what a thing is) over mechanics (how it's done). Pipes really put you [into a mechanical frame of mind](https://youtu.be/fxe5NzHDxpk?t=1438).

There's a lot going on here, and it chunks into a couple phases. This is a more typical implementation:

WIP WIP WIP switching to editor

```haskell
isExternalIPv4 ip = maybe False not maybeIsReserved
  where
    maybeIsReserved :: Maybe Bool
    maybeIsReserved = IPv4.reserved <$> normalized
    
    normalized :: Maybe IPv4
    normalized = IPv4.decode . Text.pack ipAddress

    ipAddress = extractIPFromPeerAddress $ Text.unpack ip

isExternalIPv4 ip =
  case extractIPFromPeerAddress $ Text.unpack ip
```

Reply:
Could alternately use [`Data.IPv4.isMatchedTo`](https://www.stackage.org/haddock/lts-14.7/iproute-1.7.7/Data-IP.html#v:isMatchedTo) (different library, relies on bitmasking).

Reply:
Thanks for this detailed explanation @expede 

Haskell gives us a bunch of ways to write code. Which is nice for expressivity and trying to make code read as easily as possible.

I found this one to be my prefered method
```haskell
isExternalIPv4 :: Text -> Bool
isExternalIPv4 ip = maybe False not isReserved
  where
    isReserved :: Maybe Bool
    isReserved = do
      ipAddress  <- extractIPFromPeerAddress $ Text.unpack ip
      normalized <- IPv4.decode $ Text.pack ipAddress
      return $ IPv4.reserved normalized
```

so I think I'll go with that!

Reply:
Yes, it's my favourite as well!